From 7035909a507909bafb1adc1164e5ca558c1ef13b Mon Sep 17 00:00:00 2001
From: Chad Dupuis <cdupuis@redhat.com>
Date: Thu, 21 Aug 2014 13:05:21 -0400
Subject: [ethernet] qlcnic: Do not read QLCNIC_FW_CAPABILITY_MORE_CAPS bit for 83xx adapter

Message-id: <1408626422-3941-38-git-send-email-cdupuis@redhat.com>
Patchwork-id: 88578
O-Subject: [PATCH RHEL 7.1 037/138] qlcnic: Do not read QLCNIC_FW_CAPABILITY_MORE_CAPS bit for 83xx adapter
Bugzilla: 1089352
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Shahed Shaikh <shahed.shaikh@qlogic.com>

Bug 1089352 (https://bugzilla.redhat.com/show_bug.cgi?id=1089352)

net commit id d6994ca798f5897a4342f727b21d77e01d92f093

Only 82xx adapter advertises QLCNIC_FW_CAPABILITY_MORE_CAPS bit.
Reading this bit from 83xx adapter causes the driver to skip
extra capabilities registers.

Because of this, driver was not issuing qlcnic_fw_cmd_set_drv_version()
for 83xx adapter.

This bug was introduced in commit 8af3f33db05c6d0146ad14905145a5c923770856
 ("qlcnic: Add support for 'set driver version' in 83XX").

Signed-off-by: Shahed Shaikh <shahed.shaikh@qlogic.com>
Signed-off-by: David S. Miller <davem@davemloft.net>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/net/ethernet/qlogic/qlcnic/qlcnic_83xx_hw.c b/drivers/net/ethernet/qlogic/qlcnic/qlcnic_83xx_hw.c
index 61409f9..e1405ac 100644
--- a/drivers/net/ethernet/qlogic/qlcnic/qlcnic_83xx_hw.c
+++ b/drivers/net/ethernet/qlogic/qlcnic/qlcnic_83xx_hw.c
@@ -2276,9 +2276,9 @@ int qlcnic_83xx_get_nic_info(struct qlcnic_adapter *adapter,
   temp = (cmd.rsp.arg[8] & 0x7FFE0000) >> 17;
   npar_info->max_linkspeed_reg_offset = temp;
  }
- if (npar_info->capabilities & QLCNIC_FW_CAPABILITY_MORE_CAPS)
-  memcpy(ahw->extra_capability, &cmd.rsp.arg[16],
-         sizeof(ahw->extra_capability));
+
+ memcpy(ahw->extra_capability, &cmd.rsp.arg[16],
+        sizeof(ahw->extra_capability));
 
 out:
  qlcnic_free_mbx_args(&cmd);
diff --git a/drivers/net/ethernet/qlogic/qlcnic/qlcnic_main.c b/drivers/net/ethernet/qlogic/qlcnic/qlcnic_main.c
index b8803fd..c8124a4 100644
--- a/drivers/net/ethernet/qlogic/qlcnic/qlcnic_main.c
+++ b/drivers/net/ethernet/qlogic/qlcnic/qlcnic_main.c
@@ -1133,7 +1133,10 @@ qlcnic_initialize_nic(struct qlcnic_adapter *adapter)
   if (err == -EIO)
    return err;
   adapter->ahw->extra_capability[0] = temp;
+ } else {
+  adapter->ahw->extra_capability[0] = 0;
  }
+
  adapter->ahw->max_mac_filters = nic_info.max_mac_filters;
  adapter->ahw->max_mtu = nic_info.max_mtu;
 
@@ -2162,8 +2165,7 @@ void qlcnic_set_drv_version(struct qlcnic_adapter *adapter)
  else if (qlcnic_83xx_check(adapter))
   fw_cmd = QLCNIC_CMD_83XX_SET_DRV_VER;
 
- if ((ahw->capabilities & QLCNIC_FW_CAPABILITY_MORE_CAPS) &&
-     (ahw->extra_capability[0] & QLCNIC_FW_CAPABILITY_SET_DRV_VER))
+ if (ahw->extra_capability[0] & QLCNIC_FW_CAPABILITY_SET_DRV_VER)
   qlcnic_fw_cmd_set_drv_version(adapter, fw_cmd);
 }
 
-- 
1.7.1